src/object.d

```diff
@@ -4881,3 +4881,45 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    auto nbytes = from.length * TFrom.sizeof;
+    assert((nbytes % TTo.sizeof) == 0, "array cast misalignment");
```

Comment:
This probably should be a proper error. Because if someone compiles in `-release`, this check is gone.
Also the error message should be more descriptive.

Reply:
I can't make this an error, or it won't work in `@nogc` code.  See [this comment](https://github.com/dlang/druntime/pull/2234#issuecomment-409777211) for a more detailed explanation.

Reply:
> Because if someone compiles in -release, this check is gone.

That's a good point that I hadn't considered.  Note that if the `assert` is removed, this code can no longer be considered `@trusted`.

Reply:
@wilzbach said:

> Catching Error is defined to result in undefined behavior. We don't need to cater for such code. So you can use assert?

In an e-mail to me, @WalterBright said:

> I agree with Seb. It should have been implemented that way the first time around.

So, now that I've been directed to [`staticError`](https://github.com/dlang/druntime/blob/d8a671dac21ac0a6a8b2f10ef61b50e59ba4c2c1/src/core/exception.d#L687) (Thanks @Geod24) I can throw `Error`s in `@nogc` code, but based on @WalterBright's response, I'm not sure what I should do.